ViAn/Video/overlay.cpp

```diff
@@ -118,49 +120,61 @@ void overlay::mouse_pressed(QPoint pos) {
 }
 
 /**
- * @brief overlay::mouse_pressed
+ * @brief overlay::mouse_released
  * Ends drawing on the overlay when the mouse is
  * released, if the overlay is visible.
- * @param pos coordinates
+ * @param pos Mouse coordinates.
+ * @param frame_nr Number of the frame currently shown in the video.
  */
-void overlay::mouse_released(QPoint pos) {
-    update_drawing_position(pos);
+void overlay::mouse_released(QPoint pos, int frame_nr) {
+    update_drawing_position(pos, frame_nr);
 }
 
 /**
  * @brief overlay::mouse_moved
  * Updates drawing on the overlay when the mouse is
  * moved, if the overlay is visible.
- * @param pos coordinates
+ * @param pos Mouse coordinates.
+ * @param frame_nr Number of the frame currently shown in the video.
  */
-void overlay::mouse_moved(QPoint pos) {
-    update_drawing_position(pos);
+void overlay::mouse_moved(QPoint pos, int frame_nr) {
+    update_drawing_position(pos, frame_nr);
 }
 
 /**
  * @brief overlay::update_drawing_position
  * Updates the position of the end point of the shape currently being drawn
- * @param pos
+ * @param pos Mouse coordinates.
+ * @param frame_nr Number of the frame currently shown in the video.
  */
-void overlay::update_drawing_position(QPoint pos) {
+void overlay::update_drawing_position(QPoint pos, int frame_nr) {
     if (show_overlay) {
         // The last appended shape is the one we're currently drawing.
-        drawings.last()->update_drawing_pos(pos);
+        overlays[frame_nr].last()->update_drawing_pos(pos);
     }
 }
 
 /**
  * @brief overlay::undo
- * Undo the drawings on the overlay.
+ * Undo the drawings on the overlay, if the overlay is visible.
+ * @param frame_nr Number of the frame currently shown in the video.
  */
-void overlay::undo() {
-    drawings.takeLast();
+void overlay::undo(int frame_nr) {
+    if (show_overlay) {
+        if (overlays[frame_nr].isEmpty()) {
+            return;
+        }
+        overlays[frame_nr].takeLast(); // Requires a non-empty list.
```

Comment:
Instead of return because it is empty. just negate the if statement and put the takeLast() method in the if-statement